# Предобработка данных
- OneHotEncoded
- StandardScaler

OneHotEncoded позволяет получить бинарные(0, 1) колонки(кол-во колонок = уникальные значения)
StandardScaler нормализует(приводит к одинаковым интервалам) вещественные данные

Где-то хорошо использоват StandardScaler, где-то нет

In [42]:
import pandas as pd

In [43]:
df_train = pd.read_csv('../input/titanic/train.csv')
df_test = pd.read_csv('../input/titanic/test.csv')

## Preprocessing

In [44]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [45]:
del df_train['Cabin']

In [46]:
df_train['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [47]:
df_train['Embarked'] = df_train['Embarked'].fillna('S')

In [48]:
df_train['Age'] = df_train['Age'].fillna(int(df_train['Age'].mean()))

In [49]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


In [50]:
df_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

# Обучение модели

In [18]:
# Типы данных

# Вещественные (real)
# Категориальные (categorial)
# Порядковые (categorial)

### Прямой способ

In [20]:
# Пол

In [22]:
df_train['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [24]:
df_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [25]:
is_male = []
is_female = []

In [26]:
sexes = df_train['Sex']

In [28]:
for i in range(len(sexes)):
    if sexes[i] == 'male':
        is_male.append(1)
        is_female.append(0)
    else:
        is_male.append(0)
        is_female.append(1)

In [31]:
df_train['is_male'] = is_male
df_train['is_female'] = is_female

In [33]:
# Embarked

In [36]:
df_train['Embarked'].value_counts()

S    646
C    168
Q     77
Name: Embarked, dtype: int64

In [35]:
embs = df_train['Embarked']

In [37]:
is_S = []
is_C = []
is_Q = []

In [38]:
for i in range(len(embs)):
    if embs[i] == 'S':
        is_S.append(1)
        is_Q.append(0)
        is_C.append(0)
    elif embs[i] == 'C':
        is_S.append(0)
        is_Q.append(0)
        is_C.append(1)
    else:
        is_S.append(0)
        is_Q.append(1)
        is_C.append(0)

In [40]:
df_train['emb_is_S'] = is_S
df_train['emb_is_C'] = is_C
df_train['emb_is_Q'] = is_Q

In [52]:
names = df_train['Name']

prefixes = []

for i in range(len(names)):
    name = names[i]
    pref = name.split()[1]
    
    if pref != 'Mr.' and pref != 'Miss.' and pref != 'Mrs.' and pref != 'Master.':
        pref = 'Rare'
        
    prefixes.append(pref)

df_train['Prefix'] = prefixes

### С помощью sklearn

In [73]:
# Категориальные данные

In [53]:
df_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked Prefix  
0      0         A/5 21171   7.2500        S    Mr.  
1      0          PC 17599  71.2833        C   Mrs.  
2      0  STON/O2. 3101282   7.9250        S  Miss.  
3      0            113803  53.1000        S   Mrs.  
4      0            373450   8.0500        S    Mr.

In [54]:
categorials_columns = ['Pclass', 'Sex', 'Embarked', 'Prefix']

In [61]:
X_categorial = df_train[categorials_columns]

In [59]:
from sklearn.preprocessing import OneHotEncoder

In [63]:
one_hot_encoder = OneHotEncoder()
one_hot_encoder.fit(X_categorial)

OneHotEncoder()

In [68]:
df_train['Pclass'].value_counts() #3

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [69]:
df_train['Sex'].value_counts() #2

male      577
female    314
Name: Sex, dtype: int64

In [70]:
df_train['Embarked'].value_counts() #3

S    646
C    168
Q     77
Name: Embarked, dtype: int64

In [71]:
df_train['Prefix'].value_counts() #5

Mr.        502
Miss.      179
Mrs.       121
Rare        49
Master.     40
Name: Prefix, dtype: int64

In [81]:
X_categ_one_hoted = one_hot_encoder.transform(X_categorial).toarray()

In [74]:
# Вещественные

In [75]:
df_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked Prefix  
0      0         A/5 21171   7.2500        S    Mr.  
1      0          PC 17599  71.2833        C   Mrs.  
2      0  STON/O2. 3101282   7.9250        S  Miss.  
3      0            113803  53.1000        S   Mrs.  
4      0            373450   8.0500        S    Mr.

In [76]:
real_columns = ['Age', 'SibSp', 'Parch', 'Fare']

In [77]:
X_real = df_train[real_columns]

0.2 * Age + 0.5 * SibSp + 125 * Parch + 6 * Fare = S

если результат > 120, то человек умер

0.2 * 22 + 0.5 * 1 + 0.5 * 0 + 0.2 * 7.25 = 6.3 -> выжил

In [136]:
# Прямой способ уравнения

In [133]:
min((df_train['Age'] - 0.42) / 79.58), max((df_train['Age'] - 0.42) / 79.58)

(0.0, 1.0)

In [134]:
min(df_train['SibSp'] / 8), max(df_train['SibSp'] / 8)

(0.0, 1.0)

In [129]:
min(df_train['Parch'] / 6), max(df_train['Parch'] / 6)

(0, 6)

In [135]:
min(df_train['Fare'] / 512.3292), max(df_train['Fare']  / 512.3292)

(0.0, 1.0)

In [137]:
# С помощью sklearn

In [191]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler

In [192]:
X_real.head()

Age  SibSp  Parch     Fare
0  22.0      1      0   7.2500
1  38.0      1      0  71.2833
2  26.0      0      0   7.9250
3  35.0      1      0  53.1000
4  35.0      0      0   8.0500

In [194]:
scaler = MinMaxScaler()
scaler.fit(X_real)

MinMaxScaler()

In [195]:
X_real_scaled = scaler.transform(X_real) # Все значения от -1 до 1

In [208]:
import numpy

data_train = numpy.hstack((X_real, X_categ_one_hoted))
target = df_train['Survived']

In [209]:
from sklearn.ensemble import RandomForestClassifier

In [210]:
from sklearn.model_selection import train_test_split

In [243]:
X_train, X_test, y_train, y_test = train_test_split(data_train, target, test_size=0.3)

In [244]:
model = RandomForestClassifier(n_estimators=200, max_depth=6)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

In [245]:
from sklearn.metrics import accuracy_score

In [246]:
accuracy_score(y_pred, y_test)

0.8432835820895522

In [247]:
# > данных, > метрика

In [248]:
model = RandomForestClassifier(n_estimators=200, max_depth=6)
model.fit(data_train, target)

RandomForestClassifier(max_depth=6, n_estimators=200)

# Вывод ответа

In [100]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [101]:
del df_test['Cabin']
df_test['Age'] = df_test['Age'].fillna(df_test['Age'].mean())
df_test['Fare'] = df_test['Fare'].fillna(df_test['Fare'].mean())

In [106]:
names = df_test['Name']

prefixes = []

for i in range(len(names)):
    name = names[i]
    pref = name.split()[1]
    
    if pref != 'Mr.' and pref != 'Miss.' and pref != 'Mrs.' and pref != 'Master.':
        pref = 'Rare'
        
    prefixes.append(pref)

df_test['Prefix'] = prefixes

In [107]:
df_test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Embarked Prefix  
0  34.5      0      0   330911   7.8292        Q    Mr.  
1  47.0      1      0   363272   7.0000        S   Mrs.  
2  62.0      0      0   240276   9.6875        Q    Mr.  
3  27.0      0      0   315154   8.6625        S    Mr.  
4  22.0      1      1  3101298  12.2875        S   Mrs.

In [220]:
X_test_cat = df_test[categorials_columns]
X_test_real = df_test[real_columns]
X_test_cat_one_hoted = one_hot_encoder.transform(X_test_cat).toarray()
X_test_real_scaled = scaler.transform(X_test_real)

test_data = numpy.hstack((X_test_real, X_test_cat_one_hoted))

In [221]:
answers = model.predict(test_data)

In [222]:
df_answers = pd.DataFrame()
df_answers['PassengerId'] = df_test['PassengerId']
df_answers['Survived'] = answers

In [223]:
df_answers.to_csv('answer.csv', index=False)

In [224]:
df_answers['Survived'].value_counts()

0    271
1    147
Name: Survived, dtype: int64

# Что мы не рассмотрели?


- Балансировка классов

In [253]:
df_train['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

In [255]:
# Выживших - 800, умерших - 1

In [258]:
# модель-пень, который всегда выводит 1
# accuracity_score = 800/801 = 0.9988

In [259]:
# Выживших - 800, умерших - 300

# Тогда лучше будет рассмотреть датасет из 300 выживших и 300 умерших
# Либо 500 вышивших и 500 умерших(генерируем из 300 по рандому 200 умерших)

- и многое другое